src/dune_project.ml

```diff
@@ -364,33 +397,51 @@ let parse ~dir ~lang ~packages ~file =
      in
      match
        String.Map.of_list
-         (List.map extensions ~f:(fun (e : Extension.instance) ->
-            (Syntax.name e.extension.syntax, e.loc)))
+         (List.map explicit_extensions ~f:(fun (e : Extension.instance) ->
+            (Syntax.name (Extension.syntax e.extension), e.loc)))
      with
      | Error (name, _, loc) ->
        Errors.fail loc "Extension %S specified for the second time." name
      | Ok map ->
        let project_file : Project_file.t = { file; exists = true } in
-       let extensions =
-         extensions @
+       let implicit_extensions =
          Extension.automatic ~project_file
            ~f:(fun name -> not (String.Map.mem map name))
        in
+       let extensions =
+         List.map ~f:(fun e -> (e, true)) explicit_extensions @
+         List.map ~f:(fun e -> (e, false)) implicit_extensions
+       in
        let parsing_context = make_parsing_context ~lang ~extensions in
-       let stanzas =
-         List.concat
-           (lang.data ::
-            List.map extensions ~f:(fun (ext : Extension.instance) ->
-              ext.parse_args
-                (Dsexp.Of_sexp.set_many parsing_context ext.extension.stanzas)))
+       let extension_args, extension_stanzas =
+         List.fold_left
+           extensions
+           ~init:(Univ_map.empty, [])
+           ~f:(fun (args_acc, stanzas_acc) ((instance : Extension.instance), is_explicit) ->
+             let extension = instance.extension in
+             let Extension.Extension e = extension in
+             let args =
+               let%map (arg, stanzas) = Dsexp.Of_sexp.set_many parsing_context e.stanzas in
+               let new_args_acc =
```

Comment:
It's a bit tricky to do because parse_args can only pass a `Univ_map.t`. So, that works if we return a singleton and then add it or not (that's what I just pushed). Otherwise, we need to bring make `instance` polymorphic and add an existencial. That will require new types and annotations, I'm not sure it will be better. Is there an intermediate solution I'm missing?

Reply:
Ah, that's right. Sorry, I missed that.

Reply:
Do you prefer the singleton version or the former one with `add`?

Reply:
The former one looked better to me